astroid/inference_tip.py

```diff
@@ -34,13 +26,13 @@ def clear_inference_tip_cache() -> None:
 
 
 def _inference_tip_cached(
-    func: Callable[_P, Iterator[InferenceResult]],
-) -> Callable[_P, Iterator[InferenceResult]]:
+    func: Callable[[NodeNG, InferenceContext | None], Iterator[InferenceResult]],
```

Comment:
I do think we have support for ``**kwargs``, which we should probably keep.

Reply:
Curious why you think we should keep it. Can't find any examples. Trying to be extra-cautious about plugins?

Reply:
`ImportFrom._infer` uses an additional argument. I know, I hate it as well..

Reply:
If ImportFrom is breaking the LSP  then we could design a breaking change where we don't, perhaps ?

Reply:
I'm thinking of just moving the additional argument into **kwargs for now.

Reply:
Already called by keyword:
https://github.com/pylint-dev/pylint/blob/8776ba0808f807a8915d96c0901526d7c648ec34/pylint/checkers/variables.py#L254

Reply:
Yeah but then `func` needs to be typed differently right?

Reply:
I think I'd future that. There's not much value to adding it, it just raises the number of mypy errors, and it would cause retyping all the inference functions in `node_classes.py`.

Reply:
Hmm, most of the functions in `inference.py` already accept `**kwargs` like they should. Think it is better to reflect what we want the typing to be and work towards that than adding typing of which we know it is incorrect.

Reply:
This diff
```diff
diff --git a/astroid/typing.py b/astroid/typing.py
index 85d7031d..80d585d9 100644
--- a/astroid/typing.py
+++ b/astroid/typing.py
@@ -6,6 +6,7 @@ from __future__ import annotations
 
 from typing import (
     TYPE_CHECKING,
+    Any,
     Callable,
     Generator,
     Iterator,
@@ -75,9 +76,9 @@ InferBinaryOp = Callable[
     Generator[InferenceResult, None, None],
 ]
 
-InferFn = Callable[[_NodesT, Optional["InferenceContext"]], Iterator[InferenceResult]]
+InferFn = Callable[[_NodesT, Optional["InferenceContext"], Any], Iterator[InferenceResult]]
 InferFnExplicit = Callable[
-    [_NodesT, Optional["InferenceContext"]],
+    [_NodesT, Optional["InferenceContext"], Any],
     Union[Iterator[InferenceResult], List[InferenceResult]],
 ]
 InferFnTransform = Callable[[_NodesT, InferFn], _NodesT]
```
gives mypy errors like:
```
astroid/inference_tip.py:51: error: Too few arguments  [call-arg]
astroid/inference_tip.py:57: error: Incompatible return value type (got "Callable[[NodeNG, Optional[InferenceContext], KwArg(Any)], Union[Iterator[Union[NodeNG, UninferableBase, Proxy]], List[Union[NodeNG, UninferableBase, Proxy]]]]", expected "Callable[[Any, Optional[InferenceContext], Optional[Any]], Union[Iterator[Union[NodeNG, UninferableBase, Proxy]], List[Union[NodeNG, UninferableBase, Proxy]]]]")  [return-value]
```

To avoid it, I think I'd need a `Protocol`. I take your point we don't want wrong typing, but I just thought incremental improvements were still a good thing, since I don't have time to investigate Protocol right now.

Reply:
I can look into the `Protocol`. Want me to take over? You can focus on your PR for the `_infer` thing then 😄

Reply:
Yes please!